sw/device/silicon_creator/rom/rom_isrs.S

```diff
@@ -0,0 +1,113 @@
+// Copyright lowRISC contributors (OpenTitan project).
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey_memory.h"
+#include "sw/device/lib/base/macros.h"
+#include "flash_ctrl_regs.h"
+
+.equ LOAD_ACCESS_FAULT, 5
+
+// If we get an ECC error reading from internal flash, one of these bits
+// will be set in the flash controller's FAULT_STATUS register.
+.equ PHY_ERRORS, (1 << FLASH_CTRL_FAULT_STATUS_PHY_STORAGE_ERR_BIT) | \
+                 (1 << FLASH_CTRL_FAULT_STATUS_PHY_RELBL_ERR_BIT)
+
+// NOTE: The "ax" flag below is necessary to ensure that this section
+// is allocated executable space in ROM by the linker.
+.section .rom_isrs, "ax"
+
+// -----------------------------------------------------------------------------
+
+  /**
+   * Exception handler.
+   */
+  .balign 4
+  .global rom_exception_handler
+  .type rom_exception_handler, @function
+rom_exception_handler:
+  // Save all registers to an exception stack.  The ROM locates its initial
+  // stack at `ram_end - 16` and the stack grows downwards from there, leaving
+  // 16 bytes at the top of RAM.  We need exactly 16 bytes to handle the flash
+  // exception, so we save `sp` into `mscratch` and use those 16 bytes to save
+  // the 4 registers we need to handle the flash exception.
+  csrw mscratch, sp
+  la   sp, (_stack_end - 16)
+  sw   t0,  0 * OT_WORD_SIZE(sp)
+  sw   t1,  1 * OT_WORD_SIZE(sp)
+  sw   t2,  2 * OT_WORD_SIZE(sp)
+  sw   t3,  3 * OT_WORD_SIZE(sp)
```

Comment:
@moidx In order to allow `//sw/device/silicon_creator/rom/e2e/address_translation:rom_ext_a_flash_b` to function correctly, I needed to use a reserved exception stack.

The assembly-language startup of all of our next stages initializes SP with `la sp, (stack_end-16)`.  That pseudo-instruction translates into a PC-relative address load like this:
```
auipc sp, <some constant>
addi sp, sp, <another constant>
```
This PC-relative load fails in the above-referenced test because the test setup intentionally loads the binary into the wrong location in flash, thus causing SP to have a bad value.

Reply:
This LGTM.